### PR TITLE
require disclosures and commands to be normalized

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
refactor Test

### CODE REVIEW

File: sdk/daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/CommandPreprocessor.scala [hunk 1]
```diff
@@ -220,21 +235,22 @@ private[lf] final class CommandPreprocessor(
   ): speedy.Command =
     cmd match {
       case command.ReplayCommand.Create(templateId, argument) =>
-        unsafePreprocessCreate(templateId, argument)
+        unsafePreprocessCreate(templateId, argument, strict = true)
       case command.ReplayCommand.Exercise(templateId, mbIfaceId, coid, choiceId, argument) =>
         mbIfaceId match {
           case Some(ifaceId) =>
-            unsafePreprocessExerciseInterface(ifaceId, coid, choiceId, argument)
+            discard(handleLookup(pkgInterface.lookupInterfaceInstance(ifaceId, templateId)))
```
Comment (remyhaemmerle-da) [inline sdk/daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/CommandPreprocessor.scala:242]: This was actually a bug. This is capture by one of the test.

File: sdk/daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala [hunk 2]
```diff
@@ -110,6 +122,12 @@ private[lf] final class ValueTranslator(
           case (PartyF, ValueParty(p)) =>
             SValue.SParty(p)
           case (NumericF(s), ValueNumeric(d)) =>
+            if (strict)
+              if (Numeric.scale(d) != s)
+                typeError(
+                  s"Non-normalized Numeric: the type tell us Numeric $s, but the value tell use Numeric ${Numeric
```
Comment (paulbrauner-da) [inline sdk/daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala:128]:
```suggestion
                  s"Non-normalized Numeric: the type tells us Numeric $s, but the value tells us Numeric ${Numeric
```

File: sdk/daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala [hunk 3]
```diff
@@ -181,7 +197,13 @@ private[lf] final class ValueTranslator(
                   )
               }
 
-            val oLabeledFlds = labeledRecordToMap(sourceElements).fold(typeError, identity _)
+            val oLabeledFlds = if (strict) {
+              sourceElements.iterator.collectFirst { case (Some(label), _) =>
+                typeError(s"expect no label in record but find '$label'")
```
Comment (paulbrauner-da) [inline sdk/daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala:202]:
```suggestion
                typeError(s"expected no label in record but found '$label'")
```

File: sdk/daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala [hunk 4]
```diff
@@ -193,7 +215,10 @@ private[lf] final class ValueTranslator(
               oLabeledFlds match {
                 // Not fully labelled (or reordering disabled), so order dependent
                 // Additional fields should downgrade, missing fields should upgrade
-                case None => {
+                case None =>
+                  if (strict && sourceElements.toSeq.lastOption.exists(_._2 == ValueNone))
+                    normalizationError(s"expect no trailing None")
```
Comment (paulbrauner-da) [inline sdk/daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala:220]:
```suggestion
                    normalizationError(s"expected no trailing None")
```

File: sdk/daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/preprocessing/CommandPreprocessorSpec.scala [hunk 5]
```diff
@@ -0,0 +1,515 @@
...
+
+    val IId: Ref.TypeConName = "Mod:I"
+    val IName: Ref.TypeConRef = Ref.TypeConRef(Ref.PackageRef.Name(pkg.pkgName), IId.qualifiedName)
+
+    val normalizedRecordSome = ValueRecord("", ImmArray("" -> valueParties, "" -> someSuffixedCid))
+    val normalizedRecordNone = ValueRecord("", ImmArray("" -> valueParties))
+    val nonNormalizedRecordNone = ValueRecord("", ImmArray("" -> valueParties, "" -> ValueNone))
+    val nonNormalizedRecordUpgrade =
+      ValueRecord("", ImmArray("" -> valueParties, "" -> ValueNone, "" -> ValueNone))
+    val illTypedRecord = ValueRecord("", ImmArray("" -> valueParties, "" -> ValueUnit))
```
Comment (paulbrauner-da) [inline sdk/daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/preprocessing/CommandPreprocessorSpec.scala:135]: As a special case of illTypedRecord we could also have a record that does not have enough non-optional fields.

File: sdk/daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/preprocessing/CommandPreprocessorSpec.scala [hunk 6]
```diff
@@ -0,0 +1,515 @@
...
+        validApiCreate.copy(argument = invalidUpgradeRecord) ->
+          a[Error.Preprocessing.TypeMismatch],
+        validApiCreate.copy(argument = withInvalidCidRecord) ->
+          a[Error.Preprocessing.IllegalContractId],
+        // TEST_EVIDENCE: Integrity: ill-formed exercise API command is rejected
+        validApiExe.copy(typeRef = "Mod:Undefined") ->
+          a[Error.Preprocessing.Lookup],
+        validApiExe.copy(typeRef = "Mod:S") ->
+          a[Error.Preprocessing.Lookup],
+        validApiExe.copy(choiceId = "Undefined") ->
```
Comment (paulbrauner-da) [inline sdk/daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/preprocessing/CommandPreprocessorSpec.scala:227]: We could also have a case that exercises an existing choice that doesn't belong to the template.

**Review by paulbrauner-da — APPROVED** [review]
Splendid test coverage!

File: sdk/daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/preprocessing/CommandPreprocessorSpec.scala (hunk 6 above)
Comment (remyhaemmerle-da) [reply sdk/daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/preprocessing/CommandPreprocessorSpec.scala:227]: This is the case line 225

File: sdk/daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/preprocessing/CommandPreprocessorSpec.scala (hunk 5 above)
Comment (remyhaemmerle-da) [reply sdk/daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/preprocessing/CommandPreprocessorSpec.scala:135]: This should be capture by the tests for valueTranslator.

**Pull request merged**